## Determine major work destinations within census tracts in New Jersey

##### Which tracts have the highest number of jobs that people commute to from another county within the state using LODES database
##### To understand where residents within the state are employed

* Data source (2018): https://lehd.ces.census.gov/data/
* Variables:
* Residence Census Block Code (h_geocode)
* Workplace Census Block Code (w_geocode) 
* Total number of primary jobs (S000)
* Identify top 5 work destinations where NJ residents commute to work



In [ ]:

import numpy as np
import pandas as pd
import matplotlib as mplib
import matplotlib.pyplot as plt

# For inline graphics
%matplotlib inline

In [4]:
pd.set_option('display.max_columns', None)

In [5]:
def get_nj_wac_main(year, state = "nj"):
    # state: string, two-letter code of state for which we want the data
    # year: int, the year we want to bring in data for
    
    base_url = 'https://lehd.ces.census.gov/data/lodes/LODES7/'
    file_specs = '{st}/wac/{st}_wac_SI03_JT05_{yr}.csv.gz'.format(st = state, yr = year)
    file_name = base_url + file_specs
    
    print("The URL for the file is at: " + file_name)
    output = pd.read_csv(file_name,compression='gzip')
    return(output)

In [6]:
df_wac_2018 = get_nj_wac_main(year =2018, state = 'nj')
df_wac_2018.head()

The URL for the file is at: https://lehd.ces.census.gov/data/lodes/LODES7/nj/wac/nj_wac_SI03_JT05_2018.csv.gz


w_geocode  C000  CA01  CA02  CA03  CE01  CE02  CE03  CNS01  CNS02  \
0  340010012002020   169     2   101    66     1     4   164      0      0   
1  340010024002005  1476    77   847   552     3    11  1462      0      0   
2  340010024002010     1     1     0     0     0     1     0      0      0   
3  340010024003012     8     0     4     4     0     0     8      0      0   
4  340010024003022     1     0     1     0     0     0     1      0      0   

   CNS03  CNS04  CNS05  CNS06  CNS07  CNS08  CNS09  CNS10  CNS11  CNS12  \
0      0      0      0      0      0      0      0      0      0      0   
1      0      0      0      0      0      0      0      0      0      0   
2      0      0      0      0      0      0      0      0      0      0   
3      0      0      0      0      0      0      8      0      0      0   
4      0      0      0      0      0      0      0      0      0      1   

   CNS13  CNS14  CNS15  CNS16  CNS17  CNS18  CNS19  CNS20  CR01  CR02  CR03  \
0      0      0      0      0      0      0      0    169   145    10     0   
1      0      0      0      0      0      0      0   1476  1157   121    13   
2      0      0      0      0      0      0      0      1     1     0     0   
3      0      0      0      0      0      0      0      0     7     1     0   
4      0      0      0      0      0      0      0      0     1     0     0   

   CR04  CR05  CR07  CT01  CT02  CD01  CD02  CD03  CD04  CS01  CS02  CFA01  \
0     6     0     8   162     7     0    30    51    86    95    74      0   
1   137     6    42  1405    71    15    94   308   982  1020   456      0   
2     0     0     0     1     0     0     0     0     0     1     0      0   
3     0     0     0     8     0     0     2     1     5     0     8      0   
4     0     0     0     1     0     0     0     0     1     1     0      0   

   CFA02  CFA03  CFA04  CFA05  CFS01  CFS02  CFS03  CFS04  CFS05  createdate  
0      0      0      0      0      0      0      0      0      0    20201120  
1      0      0      0      0      0      0      0      0      0    20201120  
2      0      0      0      0      0      0      0      0      0    20201120  
3      0      0      0      0      0      0      0      0      0    20201120  
4      0      0      0      0      0      0      0      0      0    20201120

In [7]:
def get_nj_rac_main(year, state = "nj"):
    # state: string, two-letter code of state for which we want the data
    # year: int, the year we want to bring in data for
    
    base_url = 'https://lehd.ces.census.gov/data/lodes/LODES7/'
    file_specs = '{st}/rac/{st}_rac_SI03_JT05_{yr}.csv.gz'.format(st = state, yr = year)
    file_name = base_url + file_specs
    
    print("The URL for the file is at: " + file_name)
    output = pd.read_csv(file_name,compression='gzip')
    return(output)

In [8]:
df_rac_2018 = get_nj_rac_main(year =2018, state = 'nj')
df_rac_2018.head()

The URL for the file is at: https://lehd.ces.census.gov/data/lodes/LODES7/nj/rac/nj_rac_SI03_JT05_2018.csv.gz


h_geocode  C000  CA01  CA02  CA03  CE01  CE02  CE03  CNS01  CNS02  \
0  340010001001009     1     1     0     0     0     0     1      0      0   
1  340010001002018     1     1     0     0     0     1     0      0      0   
2  340010002001010     1     0     1     0     0     0     1      0      0   
3  340010002002003     1     0     1     0     0     0     1      0      0   
4  340010002002012     1     0     0     1     0     0     1      0      0   

   CNS03  CNS04  CNS05  CNS06  CNS07  CNS08  CNS09  CNS10  CNS11  CNS12  \
0      0      0      0      0      0      0      0      0      0      0   
1      0      0      0      0      0      0      0      0      0      0   
2      0      0      0      0      0      0      0      0      0      0   
3      0      0      0      0      0      0      0      0      0      0   
4      0      0      0      0      0      0      0      0      0      0   

   CNS13  CNS14  CNS15  CNS16  CNS17  CNS18  CNS19  CNS20  CR01  CR02  CR03  \
0      0      0      0      0      0      0      0      1     0     0     0   
1      0      0      0      0      0      0      0      1     0     1     0   
2      0      0      0      0      0      0      0      1     0     0     0   
3      0      0      0      0      0      0      0      1     1     0     0   
4      0      0      0      0      0      0      0      1     1     0     0   

   CR04  CR05  CR07  CT01  CT02  CD01  CD02  CD03  CD04  CS01  CS02  \
0     1     0     0     1     0     0     0     0     0     0     1   
1     0     0     0     1     0     0     0     0     0     0     1   
2     1     0     0     1     0     0     0     0     1     1     0   
3     0     0     0     1     0     0     0     1     0     0     1   
4     0     0     0     1     0     0     0     0     1     0     1   

   createdate  
0    20201120  
1    20201120  
2    20201120  
3    20201120  
4    20201120

In [10]:
def get_geo_xwalk_main(state = "nj"):
    # state: string, two-letter code of state for which we want the data
    # year: int, the year we want to bring in data for
    
    base_url = 'https://lehd.ces.census.gov/data/lodes/LODES7/'
    file_specs = '{st}/{st}_xwalk.csv.gz'.format(st = state)
    file_name = base_url + file_specs
    
    print("The URL for the file is at: " + file_name)
    output = pd.read_csv(file_name,compression='gzip')
    return(output)

In [12]:
df_xwalk_2018 = get_geo_xwalk_main(state = 'nj')
df_xwalk_2018.head()

The URL for the file is at: https://lehd.ces.census.gov/data/lodes/LODES7/nj/nj_xwalk.csv.gz


/Users/NidhiTandon/opt/anaconda3/lib/python3.8/site-packages/IPython/core/interactiveshell.py:3361: DtypeWarning: Columns (27,28,29,37) have mixed types.Specify dtype option on import or set low_memory=False.
  if (await self.run_code(code, result,  async_=asy)):


tabblk2010  st stusps      stname    cty              ctyname  \
0  340090204001109  34     NJ  New Jersey  34009  Cape May County, NJ   
1  340090221011038  34     NJ  New Jersey  34009  Cape May County, NJ   
2  340210028004009  34     NJ  New Jersey  34021    Mercer County, NJ   
3  340210028004015  34     NJ  New Jersey  34021    Mercer County, NJ   
4  340210028004004  34     NJ  New Jersey  34021    Mercer County, NJ   

          trct               trctname          bgrp  \
0  34009020400     204 (Cape May, NJ)  340090204001   
1  34009022101  221.01 (Cape May, NJ)  340090221011   
2  34021002800        28 (Mercer, NJ)  340210028004   
3  34021002800        28 (Mercer, NJ)  340210028004   
4  34021002800        28 (Mercer, NJ)  340210028004   

                         bgrpname   cbsa               cbsaname  zcta  \
0     1 (Tract 204, Cape May, NJ)  36140         Ocean City, NJ  8270   
1  1 (Tract 221.01, Cape May, NJ)  36140         Ocean City, NJ  8251   
2        4 (Tract 28, Mercer, NJ)  45940  Trenton-Princeton, NJ  8629   
3        4 (Tract 28, Mercer, NJ)  45940  Trenton-Princeton, NJ  8629   
4        4 (Tract 28, Mercer, NJ)  45940  Trenton-Princeton, NJ  8629   

   zctaname    stplc stplcname     ctycsub                     ctycsubname  \
0    8270.0  9999999       NaN  3400974810   Upper township (Cape May, NJ)   
1    8251.0  9999999       NaN  3400945810  Middle township (Cape May, NJ)   
2    8629.0  9999999       NaN  3402129310  Hamilton township (Mercer, NJ)   
3    8629.0  9999999       NaN  3402129310  Hamilton township (Mercer, NJ)   
4    8629.0  9999999       NaN  3402129310  Hamilton township (Mercer, NJ)   

   stcd116 stcd116name  stsldl                        stsldlname  stsldu  \
0     3402       NJ-02   34001   General Assembly District 1, NJ   34001   
1     3402       NJ-02   34001   General Assembly District 1, NJ   34001   
2     3404       NJ-04   34014  General Assembly District 14, NJ   34014   
3     3404       NJ-04   34014  General Assembly District 14, NJ   34014   
4     3404       NJ-04   34014  General Assembly District 14, NJ   34014   

                     stslduname  stschool  \
0   State Senate District 1, NJ   3416650   
1   State Senate District 1, NJ   3410020   
2  State Senate District 14, NJ   3406540   
3  State Senate District 14, NJ   3406540   
4  State Senate District 14, NJ   3406540   

                            stschoolname  stsecon stseconname   trib tribname  \
0     Upper Township School District, NJ  9999999         NaN  99999      NaN   
1    Middle Township School District, NJ  9999999         NaN  99999      NaN   
2  Hamilton Township School District, NJ  9999999         NaN  99999      NaN   
3  Hamilton Township School District, NJ  9999999         NaN  99999      NaN   
4  Hamilton Township School District, NJ  9999999         NaN  99999      NaN   

      tsub  tsubname   stanrc  stanrcname  necta  nectaname  mil milname  \
0  9999999       NaN  9999999         NaN  99999        NaN  NaN     NaN   
1  9999999       NaN  9999999         NaN  99999        NaN  NaN     NaN   
2  9999999       NaN  9999999         NaN  99999        NaN  NaN     NaN   
3  9999999       NaN  9999999         NaN  99999        NaN  NaN     NaN   
4  9999999       NaN  9999999         NaN  99999        NaN  NaN     NaN   

      stwib                 stwibname   blklatdd   blklondd  createdate  
0  34340001  01 Atlantic-Cape May WIB  39.274472 -74.829342    20201120  
1  34340001  01 Atlantic-Cape May WIB  39.044120 -74.920564    20201120  
2  34340010             10 Mercer WIB  40.224083 -74.724779    20201120  
3  34340010             10 Mercer WIB  40.222765 -74.725446    20201120  
4  34340010             10 Mercer WIB  40.225823 -74.727454    20201120

In [13]:
def get_od_main(year, state = "nj"):
    # state: string, two-letter code of state for which we want the data
    # year: int, the year we want to bring in data for
    
    base_url = 'https://lehd.ces.census.gov/data/lodes/LODES7/'
    file_specs = '{st}/od/{st}_od_main_JT05_{yr}.csv.gz'.format(st = state, yr = year)
    file_name = base_url + file_specs
    
    print("The URL for the file is at: " + file_name)
    output = pd.read_csv(file_name,compression='gzip')
    return(output)

In [14]:
df_2018 = get_od_main(year =2018, state = 'nj')
df_2018.head()

The URL for the file is at: https://lehd.ces.census.gov/data/lodes/LODES7/nj/od/nj_od_main_JT05_2018.csv.gz


w_geocode        h_geocode  S000  SA01  SA02  SA03  SE01  SE02  SE03  \
0  340010012002020  340010001002018     1     1     0     0     0     1     0   
1  340010012002020  340010005001002     1     0     1     0     0     0     1   
2  340010012002020  340010005001003     1     0     1     0     0     0     1   
3  340010012002020  340010101022013     1     0     1     0     0     0     1   
4  340010012002020  340010102002023     1     0     0     1     0     0     1   

   SI01  SI02  SI03  createdate  
0     0     0     1    20201120  
1     0     0     1    20201120  
2     0     0     1    20201120  
3     0     0     1    20201120  
4     0     0     1    20201120

In [15]:
# Inner join on crosswalk data and O-D data 
df_final = pd.DataFrame({
    "tabblk2010": df_xwalk_2018.tabblk2010,
    "w_geocode": df_2018.w_geocode,
    "h_geocode": df_2018.h_geocode,
    "S000": df_2018.S000,
    "C000": df_rac_2018.C000,
    "trct": df_xwalk_2018.trct,
    "cty": df_xwalk_2018.cty
})
df_final 

tabblk2010     w_geocode     h_geocode  S000  C000         trct  \
0       340090204001109  3.400100e+14  3.400100e+14   1.0   1.0  34009020400   
1       340090221011038  3.400100e+14  3.400100e+14   1.0   1.0  34009022101   
2       340210028004009  3.400100e+14  3.400100e+14   1.0   1.0  34021002800   
3       340210028004015  3.400100e+14  3.400101e+14   1.0   1.0  34021002800   
4       340210028004004  3.400100e+14  3.400101e+14   1.0   1.0  34021002800   
...                 ...           ...           ...   ...   ...          ...   
169583  340299801001017           NaN           NaN   NaN   NaN  34029980100   
169584  340339900000009           NaN           NaN   NaN   NaN  34033990000   
169585  340339900000007           NaN           NaN   NaN   NaN  34033990000   
169586  340339900000004           NaN           NaN   NaN   NaN  34033990000   
169587  340339900000006           NaN           NaN   NaN   NaN  34033990000   

          cty  
0       34009  
1       34009  
2       34021  
3       34021  
4       34021  
...       ...  
169583  34029  
169584  34033  
169585  34033  
169586  34033  
169587  34033  

[169588 rows x 7 columns]

In [19]:
#pd.set_option('display.float_format', lambda x: '%.2f' % x)

In [27]:
df_final.dropna(inplace=True)

In [28]:
df_final['w_geocode'] = df_final['w_geocode'].astype('int64')

In [29]:
df_final['h_geocode'] = df_final['h_geocode'].astype('int64')

In [30]:
df_final.head()

tabblk2010        w_geocode        h_geocode  S000  C000         trct  \
0  340090204001109  340010012002020  340010001002018  1.00  1.00  34009020400   
1  340090221011038  340010012002020  340010005001002  1.00  1.00  34009022101   
2  340210028004009  340010012002020  340010005001003  1.00  1.00  34021002800   
3  340210028004015  340010012002020  340010101022013  1.00  1.00  34021002800   
4  340210028004004  340010012002020  340010102002023  1.00  1.00  34021002800   

     cty  
0  34009  
1  34009  
2  34021  
3  34021  
4  34021

In [31]:
# aggregate blocks to tracts and sum S000 column
data_agg = df_final.groupby(['trct', 'cty'], as_index=False)[['S000']].sum()
data_agg

trct    cty  S000
0    34001000100  34001 34.00
1    34001010102  34001  1.00
2    34001010501  34001  1.00
3    34001010505  34001  1.00
4    34001011201  34001  1.00
..           ...    ...   ...
353  34039038300  34039 14.00
354  34041031200  34041  3.00
355  34041031302  34041  1.00
356  34041031601  34041  1.00
357  34041031900  34041  1.00

[358 rows x 3 columns]

In [33]:
# sort by S000 column descending
data_agg = data_agg.sort_values('S000', ascending=False)

In [37]:
data_agg.head(25)

trct    cty   S000
222  34019010200  34019 223.00
224  34019010400  34019 188.00
227  34019010701  34019 187.00
239  34019011600  34019 186.00
238  34019011500  34019 182.00
141  34017012700  34017 176.00
233  34019011002  34019 169.00
328  34035052700  34035 138.00
232  34019011001  34019 130.00
226  34019010600  34019 130.00
325  34035052400  34035 117.00
235  34019011301  34019 115.00
324  34035052300  34035 106.00
242  34019011900  34019 106.00
290  34027044401  34027 106.00
332  34035052904  34035  98.00
344  34035054100  34035  96.00
228  34019010702  34019  96.00
225  34019010500  34019  95.00
322  34035052201  34035  93.00
336  34035053602  34035  91.00
223  34019010300  34019  90.00
249  34021003500  34021  85.00
216  34017019900  34017  84.00
231  34019010900  34019  84.00

In [49]:
data_agg['%'] = ((data_agg['S000'] / data_agg['S000'].sum())*100).round(2).astype(str) + '%'

In [50]:
data_agg.head()

trct    cty   S000      %
222  34019010200  34019 223.00  2.49%
224  34019010400  34019 188.00   2.1%
227  34019010701  34019 187.00  2.09%
239  34019011600  34019 186.00  2.08%
238  34019011500  34019 182.00  2.03%

In [51]:
# sort by S000 column descending
data_agg = data_agg.sort_values('%', ascending=False)

In [52]:
data_agg.head(25)

trct    cty   S000      %
222  34019010200  34019 223.00  2.49%
224  34019010400  34019 188.00   2.1%
227  34019010701  34019 187.00  2.09%
239  34019011600  34019 186.00  2.08%
238  34019011500  34019 182.00  2.03%
141  34017012700  34017 176.00  1.96%
233  34019011002  34019 169.00  1.89%
328  34035052700  34035 138.00  1.54%
232  34019011001  34019 130.00  1.45%
226  34019010600  34019 130.00  1.45%
325  34035052400  34035 117.00  1.31%
235  34019011301  34019 115.00  1.28%
290  34027044401  34027 106.00  1.18%
324  34035052300  34035 106.00  1.18%
242  34019011900  34019 106.00  1.18%
332  34035052904  34035  98.00  1.09%
344  34035054100  34035  96.00  1.07%
228  34019010702  34019  96.00  1.07%
225  34019010500  34019  95.00  1.06%
322  34035052201  34035  93.00  1.04%
336  34035053602  34035  91.00  1.02%
223  34019010300  34019  90.00   1.0%
249  34021003500  34021  85.00  0.95%
216  34017019900  34017  84.00  0.94%
231  34019010900  34019  84.00  0.94%

#### County Codes for NJ

* 34001	Atlantic	NJ
* 34003	Bergen	NJ
* 34005	Burlington	NJ
* 34007	Camden	NJ
* 34009	Cape May	NJ
* 34011	Cumberland	NJ
* 34013	Essex	NJ
* 34015	Gloucester	NJ
* 34017	Hudson	NJ
* 34019	Hunterdon	NJ
* 34021	Mercer	NJ
* 34023	Middlesex	NJ
* 34025	Monmouth	NJ
* 34027	Morris	NJ
* 34029	Ocean	NJ
* 34031	Passaic	NJ
* 34033	Salem	NJ
* 34035	Somerset	NJ
* 34037	Sussex	NJ
* 34039	Union	NJ


### The top 5 county destinations where NJ residents commute to work are Hunterdon, Hudson, Somerset, Morris, and Mercer